torch/nn/modules/activation.py

```diff
@@ -669,6 +673,210 @@ def extra_repr(self):
+
+        q = q.contiguous().view(tgt_len, bsz * self.num_heads, self.head_dim).transpose(0, 1)
+        if k is not None:
+            k = k.contiguous().view(-1, bsz * self.num_heads, self.head_dim).transpose(0, 1)
+        if v is not None:
+            v = v.contiguous().view(-1, bsz * self.num_heads, self.head_dim).transpose(0, 1)
+
+        if saved_state is not None:
+            # saved states are stored with shape (bsz, num_heads, seq_len, head_dim)
+            if 'prev_key' in saved_state:
+                prev_key = saved_state['prev_key'].view(bsz * self.num_heads, -1, self.head_dim)
+                if static_kv:
+                    k = prev_key
+                else:
+                    k = torch.cat((prev_key, k), dim=1)
+            if 'prev_value' in saved_state:
+                prev_value = saved_state['prev_value'].view(bsz * self.num_heads, -1, self.head_dim)
+                if static_kv:
+                    v = prev_value
+                else:
+                    v = torch.cat((prev_value, v), dim=1)
+            saved_state['prev_key'] = k.view(bsz, self.num_heads, -1, self.head_dim)
+            saved_state['prev_value'] = v.view(bsz, self.num_heads, -1, self.head_dim)
+
+            self._set_input_buffer(incremental_state, saved_state)
+
+        src_len = k.size(1)
+
+        if key_padding_mask is not None:
+            assert key_padding_mask.size(0) == bsz
+            assert key_padding_mask.size(1) == src_len
+
+        if self.add_zero_attn:
+            src_len += 1
+            k = torch.cat([k, k.new_zeros((k.size(0), 1) + k.size()[2:])], dim=1)
+            v = torch.cat([v, v.new_zeros((v.size(0), 1) + v.size()[2:])], dim=1)
+            if attn_mask is not None:
+                attn_mask = torch.cat([attn_mask, attn_mask.new_zeros(attn_mask.size(0), 1)], dim=1)
+            if key_padding_mask is not None:
+                key_padding_mask = torch.cat(
+                    [key_padding_mask, torch.zeros(key_padding_mask.size(0), 1).type_as(key_padding_mask)], dim=1)
+
+        attn_weights = torch.bmm(q, k.transpose(1, 2))
+        assert list(attn_weights.size()) == [bsz * self.num_heads, tgt_len, src_len]
+
+        if attn_mask is not None:
+            attn_mask = attn_mask.unsqueeze(0)
+            # if self.onnx_trace:
+            #     attn_mask = attn_mask.repeat(attn_weights.size(0), 1, 1)
+            attn_weights += attn_mask
+
+        if key_padding_mask is not None:
+            attn_weights = attn_weights.view(bsz, self.num_heads, tgt_len, src_len)
+            attn_weights = attn_weights.float().masked_fill(
+                key_padding_mask.unsqueeze(1).unsqueeze(2),
+                float('-inf'),
+            ).type_as(attn_weights)  # FP16 support: cast to float and back
+            attn_weights = attn_weights.view(bsz * self.num_heads, tgt_len, src_len)
+
+        attn_weights = F.softmax(attn_weights.float(), dim=-1).type_as(attn_weights)
```

Comment:
The native FP32 softmax is preferred now:
```F.softmax(attn_weights, dim=-1, dtype=torch.float32).type_as(attn_weights)```

Reply:
dtype=torch.float32 only if attn_weights.dtype != torch.float64

Reply:
Ah, good call. Or is it better to check for float16 explicitly?

Or maybe leave all this to apex? I'm still holding out for an apex.convert_ops_to_fp32 context manager...

Reply:
i think the dtype line can safely be:

```
dtype=torch.float32 if attn_weights.dtype == torch.float16 else torch.float32
```
```

Reply:
Thanks Soumith. I add it.

Reply:
In fact, even this 
```
F.softmax(attn_weights, dim=-1, dtype=torch.float32).type_as(attn_weights)
``` 
is not needed. What's happening with this line is softmax is computed in fp32, output in fp32 and immediately truncated to fp16 if attn_weights were fp16, no-op otherwise (I'm ignoring double here on purpose, there should be no type casts in double). 
```
F.softmax(attn_weights, dim=-1)
```
would do exact same thing. `dtype=torch.float32` output type for logsoftmax/softmax should be used only if you intend to do something with fp32 output (e.g. compute loss), not if you intend to cast it back to narrower type right away.